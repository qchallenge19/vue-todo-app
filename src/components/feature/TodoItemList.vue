<template>
    <div class="main">
        <div class="container">
            <div class="add-task-container">
                <el-row :gutter="20">
                    <el-col :xs="16" :sm="18" :md="20" :lg="21"><div>
                        <el-input v-model="newTodo" @keyup.enter="addTodo"></el-input>
                    </div></el-col>
                    <el-col :xs="8" :sm="6" :md="4" :lg="3"><div>
                        <el-button type="primary" @click="addTodo" :disabled="newTodo.length === 0" style="width: 100%">Add</el-button></div>
                    </el-col>
                </el-row>
            </div>
            <div>
                <el-row>
                    <el-col :span="24">
                        <div>
                            <el-table
                                    :data="tableData.filter(data => !search || data.description.toLowerCase().includes(search.toLowerCase()))"
                                    empty-text="No items" row-key="id" @select="handleTodoUpdate" ref="todoTable">
                                <el-table-column type="selection" style="width: 35px">
                                    <!-- eslint-disable-next-line vue/no-unused-vars -->
                                    <template slot="header" slot-scope="header">
                                    </template>
                                </el-table-column>
                                <el-table-column prop="description" min-width="200" style="text-align: left">
                                    <!-- eslint-disable-next-line vue/no-unused-vars -->
                                    <template slot="header" slot-scope="header">
                                    </template>
                                </el-table-column>
                                <el-table-column
                                        align="right">
                                    <!-- eslint-disable-next-line vue/no-unused-vars -->
                                    <template slot="header" slot-scope="scope">
                                        <el-input
                                                v-model="search"
                                                placeholder="Search"/>
                                    </template>
                                    <template slot-scope="scope">
                                        <span class="destroy" @click="handleDelete(tableData[scope.$index])"></span>
                                    </template>
                                </el-table-column>
                            </el-table>
                        </div></el-col>
                </el-row>
            </div>
        </div>
    </div>
</template>

<script>
    export default {
        name: 'todo-item-list',
        data() {
            return {
                newTodo: '',
                search: '',
            }
        },
        created () {
            this.$store.dispatch('loadTodos');
        },
        updated () {
            let todos = this.$store.getters.todos;
            todos.forEach(todo => {
                this.$refs.todoTable.toggleRowSelection(todo, todo.completed);
            });
        },
        computed: {
            tableData () {
                return this.$store.getters.todos;
            }
        },
        methods: {
            /**
             * to delete item from the todo list
             */
            handleDelete(row) {
                this.$store.dispatch('deleteTodo', row)
            },
            /**
             * to add item to the todo list
             */
            addTodo () {
                this.$store.dispatch('addTodo', this.newTodo);
                this.newTodo = '';
            },
            /**
             * to handle status update of the todo item
             * @param selection
             * @param todo
             */
            handleTodoUpdate(selection, todo) {
                this.$store.dispatch('updateTodo', todo);
            }
        },
    }
</script>
<style lang="scss">
$col-white: #fff;
$col-cornflower-blue: #a0cfff;
$col-dodger-blue: #409eff;
$col-black: #000;

.add-task-container {
  margin: auto;
  width: 90%;
}

.main {
  margin-top: 20px;
}

.el-table {
  border-bottom-left-radius: 5px;
  border-bottom-right-radius: 5px;
  text-align: center;
  width: 100%;
}

.container {
  background: $col-white;
  border: 1px solid $col-cornflower-blue;
  border-radius: 5px;
  color: $col-black;
  float: none;
  margin: auto;
  padding-top: 25px;
  width: 90%;
}

.destroy {
  align-content: center;
  bottom: 0;
  color: $col-cornflower-blue;
  display: none;
  font-size: 30px;
  height: 40px;
  margin: auto 0;
  position: absolute;
  right: 15px;
  top: 15px;
  transition: color .2s ease-out;
  width: 40px;
}

.destroy:hover {
  color: #409EFF;
}

.destroy:after {
  content: 'Ã—';
}

.el-table_2_column_4.el-table-column--selection.is-leaf .cell {
  display: none;
}

.el-table_1_column_1.el-table-column--selection.is-leaf .cell {
  display: none;
}

tr:hover .destroy {
  display: block;
}
</style>
