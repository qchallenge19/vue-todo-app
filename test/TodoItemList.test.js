import {createLocalVue, mount} from '@vue/test-utils'
import TodoItemList from '../src/components/feature/TodoItemList';
import Vuex from 'vuex';
import ElementUI from 'element-ui';

const localVue = createLocalVue();
localVue.use(Vuex);
localVue.use(ElementUI);
/* eslint-env jest */
describe('TodoItemList.vue', () => {
    let actions;
    let store;
    let getters;
    let state;
    let wrapper;

    beforeEach(() => {
        state = {
            todos: []
        };
        actions = {
            loadTodos: jest.fn(),
            addTodo: jest.fn(),
            updateTodo: jest.fn(),
            deleteTodo: jest.fn()
        };
        getters = {
            todos: () => [],
        };
        store = new Vuex.Store({
            state,
            actions,
            getters
        });

        wrapper = mount(TodoItemList, {
            sync: false,
            localVue,
            store
        });
    });
    test('renders correctly', () => {
        expect(wrapper.element).toMatchSnapshot();
    });

    test('is a TodoListItem instance', () => {
        expect(wrapper.isVueInstance()).toBeTruthy();
    });

    test('should check status of add button being disabled for empty input', () => {
        const button = wrapper.find('button');
        expect(button.attributes().disabled).toBeTruthy();
    });

    test('should call store action "addTodo" and "loadTodos" on `Add` button click', () => {
        const button = wrapper.find('button');
        button.element.disabled = false;
        button.trigger('click');
        expect(actions.addTodo).toHaveBeenCalled();
        expect(actions.loadTodos).toHaveBeenCalled();
    });
});
