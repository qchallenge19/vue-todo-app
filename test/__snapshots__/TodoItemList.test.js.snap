// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoItemList.vue renders correctly 1`] = `
<div
  class="todo"
>
  <div
    class="todo__container"
  >
    <div
      class="todo__add-task"
    >
      <div
        class="el-row"
        style="margin-left: -10px; margin-right: -10px;"
      >
        <div
          class="el-col el-col-24 el-col-xs-16 el-col-sm-18 el-col-md-20 el-col-lg-21"
          style="padding-left: 10px; padding-right: 10px;"
        >
          <div>
            <div
              class="el-input"
            >
              <!---->
              <input
                autocomplete="off"
                class="el-input__inner"
                type="text"
              />
              <!---->
              <!---->
              <!---->
              <!---->
            </div>
          </div>
        </div>
         
        <div
          class="el-col el-col-24 el-col-xs-8 el-col-sm-6 el-col-md-4 el-col-lg-3"
          style="padding-left: 10px; padding-right: 10px;"
        >
          <div>
            <div
              class="el-row"
              style="margin-left: -6px; margin-right: -6px;"
            >
              <div
                class="el-col el-col-24"
                style="padding-left: 6px; padding-right: 6px;"
              >
                <button
                  class="el-button app-button el-button--primary is-disabled"
                  disabled="disabled"
                  type="button"
                >
                  <!---->
                  <!---->
                  <span>
                    Add
                  </span>
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
     
    <div>
      <div
        class="el-row"
      >
        <div
          class="el-col el-col-24"
        >
          <div>
            <div>
              <div
                class="el-table app-table el-table--fit el-table--scrollable-x el-table--enable-row-hover"
              >
                <div
                  class="hidden-columns"
                >
                  <div />
                   
                  <div />
                   
                  <div />
                </div>
                <div
                  class="el-table__header-wrapper"
                >
                  <table
                    border="0"
                    cellpadding="0"
                    cellspacing="0"
                    class="el-table__header"
                    style="width: 315px;"
                  >
                    <colgroup>
                      <col
                        name="el-table_1_column_1"
                        width="35"
                      />
                      <col
                        name="el-table_1_column_2"
                        width="200"
                      />
                      <col
                        name="el-table_1_column_3"
                        width="80"
                      />
                    </colgroup>
                    <thead
                      class=""
                    >
                      <tr
                        class=""
                      >
                        <th
                          class="el-table_1_column_1   el-table-column--selection  is-leaf"
                          colspan="1"
                          rowspan="1"
                        >
                          <div
                            class="cell"
                          >
                            <label
                              aria-disabled="true"
                              class="el-checkbox is-disabled"
                              role="checkbox"
                            >
                              <span
                                aria-checked="mixed"
                                class="el-checkbox__input is-disabled"
                              >
                                <span
                                  class="el-checkbox__inner"
                                />
                                <input
                                  aria-hidden="true"
                                  class="el-checkbox__original"
                                  disabled="disabled"
                                  type="checkbox"
                                  value=""
                                />
                              </span>
                              <!---->
                            </label>
                          </div>
                        </th>
                        <th
                          class="el-table_1_column_2  is-left   is-leaf"
                          colspan="1"
                          rowspan="1"
                        >
                          <div
                            class="cell"
                          />
                        </th>
                        <th
                          class="el-table_1_column_3  is-right   is-leaf"
                          colspan="1"
                          rowspan="1"
                        >
                          <div
                            class="cell"
                          >
                            <div
                              class="el-input"
                            >
                              <!---->
                              <input
                                autocomplete="off"
                                class="el-input__inner"
                                placeholder="Search"
                                type="text"
                              />
                              <!---->
                              <!---->
                              <!---->
                              <!---->
                            </div>
                          </div>
                        </th>
                      </tr>
                    </thead>
                  </table>
                </div>
                <div
                  class="el-table__body-wrapper is-scrolling-left"
                >
                  <table
                    border="0"
                    cellpadding="0"
                    cellspacing="0"
                    class="el-table__body"
                    style="width: 315px;"
                  >
                    <colgroup>
                      <col
                        name="el-table_1_column_1"
                        width="35"
                      />
                      <col
                        name="el-table_1_column_2"
                        width="200"
                      />
                      <col
                        name="el-table_1_column_3"
                        width="80"
                      />
                    </colgroup>
                    <tbody>
                      <!---->
                    </tbody>
                  </table>
                  <div
                    class="el-table__empty-block"
                    style="width: 315px;"
                  >
                    <span
                      class="el-table__empty-text"
                    >
                      No items
                    </span>
                  </div>
                  <!---->
                </div>
                <!---->
                <!---->
                <!---->
                <!---->
                <div
                  class="el-table__column-resize-proxy"
                  style="display: none;"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
